File: vw_rust/src/render_pass/subpass.rs
use super::attachment::Attachment;
use crate::sys::bindings::{
    self, vw_Subpass, VwAttachmentSubpass, VwImageLayout, VwSubpassCreateArguments,
};

pub struct Subpass {
    ptr: *mut vw_Subpass,
}

pub struct SubpassBuilder {
    attachments: Vec<(Attachment, VwImageLayout)>,
}

impl SubpassBuilder {
    pub fn new() -> SubpassBuilder {
        SubpassBuilder {
            attachments: vec![],
        }
    }

    pub fn add_color_attachment(
        mut self,
        attachment: Attachment,
        layout: VwImageLayout,
    ) -> SubpassBuilder {
        self.attachments.push((attachment, layout));
        self
    }

    pub fn build(self) -> Subpass {
        let c_attachments: Vec<_> = self
            .attachments
            .iter()
            .map(|(attachment, layout)| VwAttachmentSubpass {
                attachment: attachment.to_raw(),
                currentLayout: *layout,
            })
            .collect();

        let arguments = VwSubpassCreateArguments {
            attachments: c_attachments.as_ptr(),
            attachment_count: c_attachments.len() as i32,
        };

        unsafe {
            Subpass {
                ptr: bindings::vw_create_subpass(&arguments),
            }
        }
    }
}

impl Subpass {
    pub fn as_mut_ptr(&self) -> *mut vw_Subpass {
        self.ptr
    }
}

impl Drop for Subpass {
    fn drop(&mut self) {
        unsafe {
            bindings::vw_destroy_subpass(self.ptr);
        }
    }
}

File: vw_py/RenderPass/Subpass.py
import bindings_vw_py as bindings

class Subpass:
    def __init__(self, subpass):
        self.subpass = subpass

    def __del__(self):
        print("Remove Subpass")
        bindings.vw_destroy_subpass(self.subpass)

class SubpassBuilder:
    def __init__(self):
        self.attachment_subpasses = bindings.AttachmentSubpassVector()

    def add_color_attachment(self, attachment, layout):
        attachment_subpass = bindings.VwAttachmentSubpass()
        attachment_subpass.attachment = attachment.attachment
        attachment_subpass.currentLayout = layout
        self.attachment_subpasses.push_back(attachment_subpass)
        return self
    
    def build(self):
        arguments = bindings.VwSubpassCreateArguments()
        arguments.attachments = bindings.attachment_subpass_vector_data(self.attachment_subpasses)
        arguments.attachment_count = len(self.attachment_subpasses)
        subpass = bindings.vw_create_subpass(arguments)
        return Subpass(subpass)